### ISSUE TITLE
JITServer connection status is unclear 

### ISSUE THREAD

**Issue by bconey** [eclipse-openj9/openj9#16338]
Deployed an OpenLiberty application on OCP 4.10 environment - 4 replica pods created and all connected to a JITServer pod also running on the same OCP 4.10 environment.

Began applying load (stress) to the application and at some point, 2 of the application pods displayed entries related to JITServer in the logs which indicate an error and then loss of connection to the JITServer.  

```
Pod 1 - 

139745560807168:error:140940F4:SSL routines:ssl3_read_bytes:unexpected message:ssl/record/rec_layer_s3.c:1477:
#JITServer: t=865045 Lost connection to the server (serverUID=2529813496315317418)


Pod 2 - 
139734414575360:error:140940F4:SSL routines:ssl3_read_bytes:unexpected message:ssl/record/rec_layer_s3.c:1477:
#JITServer: t=876710 Lost connection to the server (serverUID=2529813496315317418)
```

And the JITServer log contained 

```
JITServer is ready to accept incoming requests
#JITServer: t= 27637 A new client (clientUID=891492109913437483) connected. Server allocated a new client session.
#JITServer: t= 30318 A new client (clientUID=1631718612611107299) connected. Server allocated a new client session.
#JITServer: t= 32157 A new client (clientUID=411721873811045937) connected. Server allocated a new client session.
#JITServer: t= 36228 A new client (clientUID=8992861825100262930) connected. Server allocated a new client session.
139825751693056:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:ssl/record/ssl3_record.c:355:
139825751693056:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:ssl/record/ssl3_record.c:355:
```

Is the connection to the JITServer lost and then established again for those 2 pods or did it remain lost?  The JITServer log only indicates that there were errors, but does not indicate that the connection was lost at all, however the application pod logs clearly say the connection was lost but it does not indicate that it was recovered.  If it was established again, should there be some indication of that in the application pod logs?

**Comment by pshipton** [comment]
@mpirvu

**Comment by bconey** [comment]
Some more information .....

Just re-ran the scenario again and this time there was some error logged in the JITServer pod log, but in the two application pods, it did indicate a lost connection but immediately logged a message showing it connected.  So unsure if the type of connection issue determines whether a connection is reestablished.

In the JITServer pod log for this trial, it shows

```
JITServer is ready to accept incoming requests
#JITServer: t= 33456 A new client (clientUID=4231195952000416183) connected. Server allocated a new client session.
#JITServer: t= 35562 A new client (clientUID=6458182215194516952) connected. Server allocated a new client session.
#JITServer: t= 36835 A new client (clientUID=16889239032530771592) connected. Server allocated a new client session.
#JITServer: t= 37490 A new client (clientUID=18245426256939221094) connected. Server allocated a new client session.
139943223396096:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl/record/ssl3_record.c:677:
139943205566208:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl/record/ssl3_record.c:677:
```

and for the pods that it seems to have affected the following is logged - note that I did not see any type of error message in the application log prior to the lost connection message this time only those in the JITServer log :

```
Pod #1:
#JITServer: t=297370 Lost connection to the server (serverUID=3875376080527049033)
#JITServer: t=298592 Connected to a server (serverUID=3875376080527049033)

Pod #2:
#JITServer: t=293898 Lost connection to the server (serverUID=3875376080527049033)
#JITServer: t=295189 Connected to a server (serverUID=3875376080527049033)

```

**Comment by mpirvu** [comment]
@bconey 
Just for clarification, at the clients, the message "Lost connection to the server" was not preceded by a message of the form "Connected to a server" ?

Could you please add the following options `-Xjit:verbose={compilePerformance},verbose={JITServer}`  both at the server and client, repeat the experiment and send me the full logs?

**Comment by mpirvu** [comment]
I had experiment where the server was started after the client. The relevant messages are:
```
#JITServer: t=    10 Lost connection to the server (serverUID=0). Retry immediately: 0.
...
#JITServer: t=  7140 Connected to a server (serverUID=13301308956792013569)
```
So, the client attempted to connect to the server, but couldn't. I guess the first message should be more explicit in that a successful connection was never established at that point.
When the server become available, the client connected successfully, hence the second message.

**Comment by bconey** [comment]
@mpirvu Sorry for the confusion.  In both these situations when the application first started, at the top of the log was a message like `#JITServer: t= 60 Connected to a server (serverUID=<serverid>)`, that showed a successful connection to the JITServer.  After that was some normal application logging and then sometime after the load was started is when the `#JITServer: t=297370 Lost connection to the server (serverUID=<serverid>)` message was logged.  

In the original case, that "Lost connection" message was preceded by an error message like  `139734414575360:error:140940F4:SSL routines: ....`.

**Comment by mpirvu** [comment]
ok, so the client connected, then at some point experienced some errors and then (at least in one case) it connected again.
If the client experiences some networking issue (in this case it was something related to encryption) it will try to reconnect with an exponential back-off.

Could you please let me know the OpenJ9 version used here. During the summer we fixed some SSL related errors and I wonder if using a newer OpenJ9 build may make those decryption failures go away.

**Comment by bconey** [comment]
The app we are using was compiled with jdk8 but it was built with with the openj9 java17 (version 17.0.4.1+1. ?).

**Comment by bconey** [comment]
@mpirvu Do you still also need the trace previously requested and are they to just be attached to this issue when generated?

**Comment by mpirvu** [comment]
@bconey 
I checked what was delivered to 17.0.4.1 (which is OpenJ9 release 0.33.1) and the fixes I mentioned above are not included.
Could you please try the newest OpenJ9 release (0.35.0) in your tests and see if the same decryption error still happens?
This is what "java version" shows for release 0.35.0
```
IBM Semeru Runtime Open Edition 17.0.5.0 (build 17.0.5+8)
Eclipse OpenJ9 VM 17.0.5.0 (build openj9-0.35.0, JRE 17 Linux amd64-64-Bit Compressed References 20221018_325 (JIT enabled, AOT enabled)
OpenJ9   - e04a7f6c1
OMR      - 85a21674f
JCL      - 32d2c409a33 based on jdk-17.0.5+8)
```

If not, then, yes I would like to see the entire trace which can be zipped and attached to this issue (unless there is something confidential from Liberty point of view).

**Comment by bconey** [comment]
@mpirvu Not certain if we can try with that release - the tester who actually owns the building of the app is out until next week; will review upon their return.  In the meantime, hopefully the recreate was successful and proper tracing caught.  

The attached daytrader7-clientlogs.zip contains the logs from the 4 liberty server pods that were running.  Due to the way in which we had to set the trace, our initial JITServer was stopped and a new one started, so all of these servers initially connected to one JITServer and then lost connection to it and then connected to a new JITServer.  It looks like only one of the servers after connecting to the new JITServer lost connection and it doesn't look like it restarted as seen in the daytrader7-5bdb6d5df7-lh9j7.vlog.txt; see around line 17675 for the statement for the connection that is lost but doesn't seem to get re-established (the previous lost connection at line 17614 was from the old JITServer going down - and the new connection with the new JITServer happens around line 17661 ...  I think).  

The attached daytrader7-semerucomplog.zip contains just the log file for the semeru-complier pod (this is the new JITServer that started up after setting the trace).  There around line 245 is where I believe the error that might be related to that lost connection, but it is unclear since I don't see any way to link it back to a particular client but hopefully you can.

[daytrader7-clientlogs.zip](https://github.com/eclipse-openj9/openj9/files/10059466/daytrader7-clientlogs.zip)
[daytrader7-semerucomplog.zip](https://github.com/eclipse-openj9/openj9/files/10059467/daytrader7-semerucomplog.zip)

**Comment by bconey** [comment]
Here are another set of traces that a teammate took but I am not certain of the scenario.  There is a jittrace-server.zip file that contains trace and log from the JITServer pod and a jittrace-client.zip file that contains traces from 2 servers that were running for that scenario.  

In the jittrace-server.zip there is a semeru-compiler-pod.log that shows the SSL errors generated were most like those in the scenario I had where I saw them generated and the connection to the JITServer was lost by the servers for a short while, but then they reconnected.  

[jittrace-server.zip](https://github.com/eclipse-openj9/openj9/files/10060339/jittrace-server.zip)
[jittrace-client.zip](https://github.com/eclipse-openj9/openj9/files/10060340/jittrace-client.zip)

**Comment by mpirvu** [comment]
For the first set of logs, indeed the server is shutdown midway and all clients lose connection and compiling locally (15, 18, 24 and 19 local compilations respectively).
The first client also experiences a 'write error' which makes it perform one extra local compilation, but all is well otherwise.
The second client has an `SSL_connect` failure which makes it perform 8 more local compilations. It does reconnect and performs all the remaining compilations remotely (at the server), but there is no "Connected to a server" message. There is a message of the form "#JITServer: SSL connection on socket 0x6d, Version: TLSv1.3, Cipher: TLS_AES_256_GCM_SHA384" though.
I will look into why that "Connected to a server" message is missing.

**Comment by mpirvu** [comment]
I have reproduced the SSL error mesages with OpenJ9 0.33.1 and I verified that the errors go away once I moved up to OpenJ9 0.35.0

**Comment by tam512** [comment]
@mpirvu I think we also need to document this so that customers know which version of Liberty to update if they see the problem

**Comment by leochr** [comment]
@tam512 could you please raise an issue in https://github.ibm.com/websphere/operator-docs/issues with the details (error message, versions, etc)? Thank you.

**Comment by mpirvu** [comment]
SSL errors are fixed in OpenJ9 9.35.0
Connection messages are improved through #16394 which I am planning to port to next OpenJ9 release 0.36.0

**Comment by mpirvu** [comment]
This issue is fixed in current releases of OpenJ9, hence closing.
